Make impls of AsyncAggregateResponseBatchCursor more responsive to close signals and improve/fix releasing of resources

stIncMale opened the pull request:
It is recommended to review the PR commit by commit for the commits described below: each commit serves its own purpose, reviewing them one by one should simplify the process. We may also have a meeting for a code walkthrough.

Short explanations (what I recall) for the most important commits (see commit messages for more info):
1. [Change implementations of AsyncAggregateResponseBatchCursor to make them more responsive to close signals (4110f51)](https://github.com/mongodb/mongo-java-driver/pull/651/commits/4110f51857f760419953a1b340aed4232eca3f96) - Implements the fix expressed in [JAVA-3907 scenario 2)](https://jira.mongodb.org/browse/JAVA-3907?focusedCommentId=3577246&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-3577246). However, as a result of this change (which does not create new references to reference counted objects), the method `killCursorOnClose` stopped working because of inability to borrow a connection when running the [`Hang`](https://jira.mongodb.org/browse/JAVA-3907?focusedCommentId=3577246&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-3577246) litmus code (this definitely caused many other tests to hang). Additionally, `AsyncQueryBatchCursorSpecification.'should close cursor after getMore finishes if cursor was closed while getMore was in progress and getMore returns a response'` also started failing.
2. [Fix connection reference counting in AsyncQueryBatchCursor (f692551)](https://github.com/mongodb/mongo-java-driver/pull/651/commits/f692551d2b5eb4f7ced5dcaf003897a055c24046) - Fixes reference counting for connections. `AsyncQueryBatchCursorSpecification.'should close cursor after getMore finishes if cursor was closed while getMore was in progress and getMore returns a response'` still fails.
3. [Start releasing connectionSource in AsyncQueryBatchCursor.close (3262a2c)](https://github.com/mongodb/mongo-java-driver/pull/651/commits/3262a2c9c7df7bcdff80ace86c422469dafcc546) - Fixes `AsyncQueryBatchCursorSpecification.'should close cursor after getMore finishes if cursor was closed while getMore was in progress and getMore returns a response'` and changes `close` so that it releases `connectionSurce` synchronously.

stIncMale commented on driver-core/src/test/unit/com/mongodb/internal/operation/AsyncQueryBatchCursorSpecification.groovy:
```diff
@@ -626,7 +626,7 @@ class AsyncQueryBatchCursorSpecification extends Specification {
 
         then:
-        connectionSource.getCount() == 1
+        connectionSource.getCount() == 0
```
`cursor` releases `connectionSource` when it is closed.

stIncMale commented:
This looks like a too tight coupling between tests and the internals of the code:
`DefaultServerConnectionSpecification.should execute command protocol asynchronously`
```
3: argument == expected
   |        |  |
   |        |  <com.mongodb.internal.async.ErrorHandlingResultCallback@2f8fa6b6 wrapped=Mock for type 'SingleResultCallback' logger=Mock for type 'Logger'>
   |        false
   <com.mongodb.internal.async.ResourcePreservingCallback@44f69106 resource=com.mongodb.internal.connection.DefaultServerConnection@5f0162f1 wrapped=com.mongodb.internal.async.ErrorHandlingResultCallback@2f8fa6b6>
```

rozza commented:
From the evergreen logs resources there are still some resources being leaked: 

> com.mongodb.MongoTimeoutException: Timed out waiting for server session pool in use count to drop to 0.  Now at: 2

I was able to reproduce the error locally using the latest build.

stIncMale commented on driver-core/src/test/unit/com/mongodb/internal/connection/DefaultServerConnectionSpecification.groovy:
```diff
@@ -211,7 +213,7 @@ class DefaultServerConnectionSpecification extends Specification {
         then:
         1 * executor.executeAsync({ compare(new InsertProtocol(namespace, true, insertRequest), it) }, internalConnection,
-                callback)
+                {expect callback, isTheSameAs(it.wrapped)})
```
Had to unwrap `callback` to make the expectation of `executor.executeAsync` work as it was originally.

stIncMale commented:
The only test failed is [com.mongodb.reactivestreams.client.ChangeStreamsTest.shouldPassAllOutcomes_change-streams.json__Test_drop_and_invalidate_event_types_|https://evergreen.mongodb.com/test_log/mongo_java_driver_tests_jdk8_unsecure__version~4.4_os~linux_topology~replicaset_auth~noauth_ssl~nossl_jdk~jdk8_test_patch_9f90aa8aabe017b4e7993a1ade2333325e708e29_60148dc1d6d80a425d84b8d6_21_01_29_22_35_46/0/com.mongodb.reactivestreams.client.ChangeStreamsTest.shouldPassAllOutcomes_change-streams.json__Test_drop_and_invalidate_event_types_], and it was failing previously with the same exception: https://evergreen.mongodb.com/task/mongo_java_driver_tests_jdk_secure__version~4.0_os~linux_topology~replicaset_auth~auth_ssl~ssl_jdk~jdk11_test_017bf9a94398abf40ffbd33de51cbbb0d7e16e47_21_01_27_02_36_05

stIncMale commented:
I split this MR into two: https://github.com/mongodb/mongo-java-driver/pull/654 (JAVA-3907) and https://github.com/mongodb/mongo-java-driver/pull/655 (JAVA-3978).

The reason for splitting in addition to the separation of concerns is: I realized that release-early reference counting (the approach currently used in the codebase) is generally speaking not reconcilable with the structured approach. I explained this in detail in the commit https://github.com/mongodb/mongo-java-driver/pull/654/commits/0909eab325c7e1e80dfe65f2089606cb07bb7b28.

Despite we cannot have the structured reference counting, we still may use the idea of self-managing reference-counting. It is described in JAVA-3978 and implemented in https://github.com/mongodb/mongo-java-driver/pull/655.
